server/src/main/java/org/opensearch/index/shard/IndexShard.java

```diff
@@ -4809,6 +4809,14 @@ private boolean localDirectoryContains(Directory localDirectory, String file, lo
                 return true;
             } else {
                 logger.warn("Checksum mismatch between local and remote segment file: {}, will override local file", file);
+                // If there is a checksum mismatch and we are not serving reads it is safe to go ahead and delete the file now.
+                // Outside of engine resets this method will be invoked during recovery so this is safe.
+                if (isReadAllowed() == false) {
+                    localDirectory.deleteFile(file);
+                } else {
+                    // segment conflict with remote store while the shard is serving reads.
+                    failShard("Local copy of segment " + file + " has a different checksum than the version in remote store", null);
```

Comment:
are we affecting the availability here by failing the shard in case the directory implementation was going to replace the local file?

Reply:
1. With node-node segment replication, when there is difference in store metadata b/w primary and replica, we fail the replica shard. With this change, we allow the file to be overridden when it serves read. Why do we have special handling for remote store ? Is checksum mis-match state a norm under some edge-case scenario.
2. With refactor work in https://github.com/opensearch-project/OpenSearch/pull/8767 we are moving the file download logic inside `RemoteStoreReplicationSource` and diff calculation inside `SegmentReplicationTarget`. With this, we may not need to change this method considering this change is intended only for round of segment replication (I understand it is based on failing IT).

Reply:
@linuxpi - The idea here is to treat this as a corruption case and reset safely to the source of truth which is the remote store.  So yes this could cause an availability hit, the alternative is deleting the file manually while the engine is active and serving reads which I think is more dangerous and would result in corruption elsewhere.

With this flow the only time this should ever hit is if we are doing an engine reset with `resetEngineToGlobalCheckpoint` upon failover.  During that method the new primary syncs with the remote store to pick up where the previous primary left off.  If the replica has a conflicting segment from somewhere we need to treat it as a corruption case.  ll other invocations here the shard is not yet recovered (peer/store recovery).

This is also why it is critical for us to perform cksum validation on upload/download and to introduce features like point in time restore so we can catch and react to bit flips etc in the store.

@dreamer-89 After your refactor this method is still used with segrep in two places, during recovery and when opening/resetting the engine.  What I am proposing is we handle this case similarly in that we fail the shard if it is active & serving reads but overwrite if the shard is still recovering.

Reply:
I agree with the changes here.